KMeans.java

```diff
@@ -0,0 +1,126 @@
+import java.io.*;
+import java.util.*;
+import java.lang.Math; 
+
+public class KMeans {
+
+	private Vector midpoint (List<Vector> listofvectors) { //computes + returns a midpoint Vector from a given list of Vectors
+		
+		List<Double> mp_coordinates = new ArrayList<Double>();
+		double number_of_vectors = listofvectors.size();
+		double length_of_vector = listofvectors.get(0).numbers.size();	
+			
+		for (int dimension = 0; dimension < length_of_vector; dimension ++) {
+			
+			double coordinate = 0;
+
+			for (int curr = 0; curr < number_of_vectors; curr ++) {
+				
+				coordinate += listofvectors.get(curr).numbers.get(dimension);
+			}
+				
+			coordinate /= number_of_vectors;
+			mp_coordinates.add(coordinate);
+				
+		}
+		return new Vector(mp_coordinates);
+		
+	}
+	
+	private int closestCenter (Vector start_vector, List<Vector> listofcenters) { // given a starting vector and list of cluster centers, returns closest center's index
+		
+		int closest_index = 0;
+		double distance_to_closest = start_vector.distance(listofcenters.get(0));
+		
+		
+		for (int curr = 1; curr < listofcenters.size(); curr++) {
+			double distance = start_vector.distance(listofcenters.get(curr));
+			
+			if (distance < distance_to_closest) {
+				closest_index = curr;
+				distance_to_closest = distance;
+			}
+		}
+		
+		return closest_index;
+	}
+	
+	private boolean is_clustering_finished (List<Vector> old_centers, List<Vector> current_centers) { // computes normalized difference between cluster centers of 2 subsequent iterations [for all clusters], returns whether clustering is finished
+	
+		for (int index = 0; index < current_centers.size(); index++) {
+			double distance_between_centers = old_centers.get(index).distance(current_centers.get(index));
+			double magnitude = current_centers.get(index).magnitude();
+			double normalized_distance = distance_between_centers / magnitude;
+
+			
+			if (normalized_distance > 0.00000001) {
```

Comment:
Constants like this should be specified in `CONSTANT_CASE` near the top of the file. (https://google.github.io/styleguide/javaguide.html#s5.2.4-constant-names)